be/src/exprs/agg/covariance.h

```diff
@@ -0,0 +1,296 @@
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#pragma once
+
+#include <cmath>
+
+#include "column/type_traits.h"
+#include "exprs/agg/aggregate.h"
+#include "types/logical_type.h"
+namespace starrocks {
+
+// aditional state for Corelation
+struct CorelationAggregateState {
+    double m2X{};
+    double m2Y{};
+};
+
+struct EmptyAggregateState {};
+
+template <bool isCorelation>
+struct CovarianceCorelationAggregateState
+        : public std::conditional_t<isCorelation, CorelationAggregateState, EmptyAggregateState> {
+    // Average value of first column.
+    double meanX{};
+    // Average value of second column.
+    double meanY{};
+
+    double c2{};
```

Comment:
Since $Conv(x,y) = E(xy) - E(x)*E(y)$  and $Variance(x) = E(x^2) - E^2(x)$, could we just record `sum_xy`, `sum_x`, `sum_y`, `sum_x_2`, `sum_y_2` instead of the mean and variance? Then we needn't calculate the mean and variance in the process.

Reply:
yes,but this fomulat is not stable, when data size is big.